official/coin_build/man/man3/ScXMLSendElt.3
.TH "ScXMLSendElt" 3 "Sun May 28 2017" "Version 4.0.0a" "Coin" \" -*- nroff -*-
.ad l
.nh
.SH NAME
ScXMLSendElt \- the \fC<\fP send> SCXML element\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Inventor/scxml/ScXMLSendElt\&.h>\fP
.PP
Inherits \fBScXMLExecutableElt\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "virtual \fBSoType\fP \fBgetTypeId\fP (void) const"
.br
.RI "Returns the type identification of an object derived from a class inheriting \fBSoBase\fP\&. This is used for run-time type checking and 'downward' casting\&. "
.ti -1c
.RI "virtual void \fBsetEventAttribute\fP (const char *event)"
.br
.ti -1c
.RI "virtual const char * \fBgetEventAttribute\fP (void) const"
.br
.ti -1c
.RI "virtual void \fBsetTargetAttribute\fP (const char *event)"
.br
.ti -1c
.RI "virtual const char * \fBgetTargetAttribute\fP (void) const"
.br
.ti -1c
.RI "virtual void \fBsetTargetTypeAttribute\fP (const char *event)"
.br
.ti -1c
.RI "virtual const char * \fBgetTargetTypeAttribute\fP (void) const"
.br
.ti -1c
.RI "virtual void \fBsetSendIDAttribute\fP (const char *event)"
.br
.ti -1c
.RI "virtual const char * \fBgetSendIDAttribute\fP (void) const"
.br
.ti -1c
.RI "virtual void \fBsetDelayAttribute\fP (const char *event)"
.br
.ti -1c
.RI "virtual const char * \fBgetDelayAttribute\fP (void) const"
.br
.ti -1c
.RI "virtual void \fBsetNameListAttribute\fP (const char *event)"
.br
.ti -1c
.RI "virtual const char * \fBgetNameListAttribute\fP (void) const"
.br
.ti -1c
.RI "virtual void \fBsetHintsAttribute\fP (const char *event)"
.br
.ti -1c
.RI "virtual const char * \fBgetHintsAttribute\fP (void) const"
.br
.ti -1c
.RI "virtual SbBool \fBhandleXMLAttributes\fP (void)"
.br
.ti -1c
.RI "virtual void \fBcopyContents\fP (const \fBScXMLElt\fP *rhs)"
.br
.ti -1c
.RI "virtual const \fBScXMLElt\fP * \fBsearch\fP (const char *attrname, const char *attrvalue) const"
.br
.ti -1c
.RI "\fBScXMLEvent\fP * \fBcreateEvent\fP (\fBScXMLEventTarget\fP *host) const"
.br
.ti -1c
.RI "virtual void \fBexecute\fP (\fBScXMLStateMachine\fP *statemachine) const"
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static \fBScXMLEltReader\fP * \fBgetElementReader\fP (void)"
.br
.ti -1c
.RI "static \fBSoType\fP \fBgetClassTypeId\fP (void)"
.br
.ti -1c
.RI "static void * \fBcreateInstance\fP (void)"
.br
.ti -1c
.RI "static void \fBinitClass\fP (void)"
.br
.ti -1c
.RI "static void \fBcleanClass\fP (void)"
.br
.in -1c
.SS "Protected Attributes"

.in +1c
.ti -1c
.RI "char * \fBevent\fP"
.br
.ti -1c
.RI "char * \fBtarget\fP"
.br
.ti -1c
.RI "char * \fBtargettype\fP"
.br
.ti -1c
.RI "char * \fBsendid\fP"
.br
.ti -1c
.RI "char * \fBdelay\fP"
.br
.ti -1c
.RI "char * \fBnamelist\fP"
.br
.ti -1c
.RI "char * \fBhints\fP"
.br
.in -1c
.SS "Additional Inherited Members"
.SH "Detailed Description"
.PP 
the \fC<\fP send> SCXML element\&. 

The \fC<send>\fP element is only useful in profiles that have an implementation of the External Communication Module\&. Both the \fC'minimum'\fP and the \fC'x-coin'\fP profile have implementations, but the \fC'minimum'\fP profile does not implement The Data Module, so only plain events without associated data can be sent\&.
.PP
The \fC<send>\fP element can take the following attributes: 
.PD 0

.IP "\(bu" 2
\fCevent\fP 
.IP "\(bu" 2
\fCtarget\fP 
.IP "\(bu" 2
\fCtargettype\fP 
.IP "\(bu" 2
\fCsendid\fP - not supported 
.IP "\(bu" 2
\fCdelay\fP 
.IP "\(bu" 2
\fCnamelist\fP 
.IP "\(bu" 2
\fChints\fP - not supported
.PP
The <send> element sends \fBexternal\fP events\&. State machines have typically two event queues - one internal, and one external\&. The internal event queue has full priority over the external event queue\&. State machines  can send external events to itself, but it is more likely internal events are what you want\&. To send internal events, see the \fC<event>\fP element implemented in \fBScXMLEventElt\fP\&. To send an external event to the containing state machine, do not specify the \fCtarget\fP attribute\&.
.PP
The \fCevent\fP attribute specifies the event name for the event that is being sent\&. Event names are dot-separated names like \fC'error\&.InvalidExpr'\fP or \fC'x-coin-navigation\&.Rotate\&.BEGIN'\fP, and the event names are treated case-sensitive in the current implementation of Coin SCXML\&. The \fCevent\fP attribute is required\&.
.PP
The \fCtargettype\fP attribute specifies the type of the target system\&. Other SCXML state machines have the type \fC'scxml'\fP\&. Services related to Coin navigation systems have the targettype \fC'x-coin-navigation'\fP\&. When sending external events to yourself, \fCtargettype\fP can be omitted\&.
.PP
The \fCtarget\fP attribute is the name of the target event system\&. If it is not specified, it means the event will be sent to the containing state machine\&. Coin navigation services are typically names \fC'Rotate'\fP, but for other SCXML state machines, the \fC<scxml>\fP \fCname\fP attribute decides this\&. That name is not unique though if multiple such names have been created, so \fCtarget\fP can also be the state machine \fC_sessionid\fP identifier\&. (FIXME: check if this is implemented)
.PP
The \fCsendid\fP attribute is not supported\&.
.PP
If \fCdelay\fP is specified, it must take the form of \fC'{number}s'\fP or \fC'{number}ms'\fP giving you the option of specifying the delay in number of seconds or milliseconds before the event should be sent\&. For delayed events to be sent, the Coin delayqueue processing must be performed\&. Most Coin users do that wether they know it or not\&. If \fCdelay\fP is not specified, the event is passed to the target immediately\&. If a state machine is deleted, any events scheduled to be sent out in a delayed manner are going to be missing in action\&. The delay queue is a feature of the sending state machine, not the receiving one\&.
.PP
The \fCnamelist\fP attribute specifies which data variables to include with the event for the event target\&. This attribute is only useful in profiles that implement the Data Module\&. The \fC'minimum'\fP profile does not, but the \fC'x-coin'\fP profile does\&. For the \fC'x-coin'\fP profile, you can specify \fC<data>\fP variables by addressing them with the \fC_data\fP\&.{ID} scheme, or you can specify temporary variables, either by using the \fCcoin:temp\fP\&.{ID} scheme or have the prefix implicit by just specifying the {ID}\&. The is the most useful method, since event targets do not want to know where in a data model a variable comes from, nor the name it had where it was contained\&. With temporary variables, the values can get names that are intended for the event target instead of the name used at the source\&.
.PP
The \fChints\fP attribute is not supported\&.
.PP
\fBSee also:\fP
.RS 4
\fBScXMLEventTarget\fP, \fBScXMLEventElt\fP, \fBScXMLDataElt\fP, \fBScXMLAssignElt\fP 
.RE
.PP

.SH "Member Function Documentation"
.PP 
.SS "\fBSoType\fP ScXMLSendElt::getTypeId (void) const\fC [virtual]\fP"

.PP
Returns the type identification of an object derived from a class inheriting \fBSoBase\fP\&. This is used for run-time type checking and 'downward' casting\&. 
.PP
\fBSee also:\fP
.RS 4
\fBSoBase::getTypeId() const\fP 
.RE
.PP

.PP
Implements \fBScXMLExecutableElt\fP\&.
.SS "SbBool ScXMLSendElt::handleXMLAttributes (void)\fC [virtual]\fP"
This method is called when the file reader has set all the XML attributes on the object, and wants the object to handle them and report if the values were ok or if a read error should be produced\&.
.PP
\fBReturns:\fP
.RS 4
TRUE if the attributes are ok, and FALSE on error\&.
.RE
.PP
This base class implementation does nothing and just returns TRUE\&. 
.PP
Reimplemented from \fBScXMLElt\fP\&.
.SS "const \fBScXMLElt\fP * ScXMLSendElt::search (const char * attrname, const char * attrvalue) const\fC [virtual]\fP"
This method searches the SCXML structure for an element with the given attributevalue for the given attribute\&.
.PP
Returns NULL if nothing was found\&.
.PP
This function needs to be reimplemented to traverse child elements\&. 
.PP
Reimplemented from \fBScXMLElt\fP\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for Coin from the source code\&.


insanity/coin/src/scxml/ScXMLSendElt.cpp
/**************************************************************************\
 * Copyright (c) Kongsberg Oil & Gas Technologies AS
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 * 
 * Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 * 
 * Redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution.
 * 
 * Neither the name of the copyright holder nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
\**************************************************************************/

#include <Inventor/scxml/ScXMLSendElt.h>

/*!
  \class ScXMLSendElt ScXMLSendElt.h Inventor/scxml/ScXMLSendElt.h
  \brief the \c &lt; send&gt; SCXML element.

  The \c &lt;send&gt; element is only useful in profiles that have
  an implementation of the External Communication Module.  Both
  the \c "minimum" and the \c "x-coin" profile have implementations,
  but the \c "minimum" profile does not implement The Data Module, so
  only plain events without associated data can be sent.

  The \c &lt;send&gt; element can take the following attributes:
  \li \c event
  \li \c target
  \li \c targettype
  \li \c sendid - not supported
  \li \c delay
  \li \c namelist
  \li \c hints - not supported

  The &lt;send&gt; element sends \b external events.  State machines have
  typically two event queues - one internal, and one external.  The internal
  event queue has full priority over the external event queue.
  State machines \i can send external events to itself, but it is more
  likely internal events are what you want.  To send internal events, see the
  \c &lt;event&gt; element implemented in ScXMLEventElt.
  To send an external event to the containing state machine, do not specify
  the \c target attribute.

  The \c event attribute specifies the event name for the event that is
  being sent.  Event names are dot-separated names like \c "error.InvalidExpr"
  or \c "x-coin-navigation.Rotate.BEGIN", and the event names are treated
  case-sensitive in the current implementation of Coin SCXML.  The \c event
  attribute is required.

  The \c targettype attribute specifies the type of the target system.
  Other SCXML state machines have the type \c "scxml".  Services related
  to Coin navigation systems have the targettype \c "x-coin-navigation".
  When sending external events to yourself, \c targettype can be omitted.

  The \c target attribute is the name of the target event system.
  If it is not specified, it means the event will be sent to the containing
  state machine.
  Coin navigation services are typically names \c "Rotate", but for other
  SCXML state machines, the \c &lt;scxml&gt; \c name attribute decides
  this.  That name is not unique though if multiple such names have been
  created, so \c target can also be the state machine \c _sessionid identifier.
  (FIXME: check if this is implemented)

  The \c sendid attribute is not supported.

  If \c delay is specified, it must take the form of \c "{number}s" or
  \c "{number}ms" giving you the option of specifying the delay in
  number of seconds or milliseconds before the event should be sent.
  For delayed events to be sent, the Coin delayqueue processing must
  be performed. Most Coin users do that wether they know it or not. If
  \c delay is not specified, the event is passed to the target
  immediately. If a state machine is deleted, any events scheduled to
  be sent out in a delayed manner are going to be missing in action.
  The delay queue is a feature of the sending state machine, not the
  receiving one.

  The \c namelist attribute specifies which data variables to include
  with the event for the event target.  This attribute is only useful
  in profiles that implement the Data Module.  The \c "minimum" profile does
  not, but the \c "x-coin" profile does.  For the \c "x-coin" profile,
  you can specify \c &lt;data&gt; variables by addressing them with the
  \c _data.{ID} scheme, or you can specify temporary variables, either by
  using the \c coin:temp.{ID} scheme or have the prefix implicit by just
  specifying the {ID}.  The is the most useful method, since event targets
  do not want to know where in a data model a variable comes from, nor the
  name it had where it was contained.  With temporary variables, the values
  can get names that are intended for the event target instead of the name
  used at the source.

  The \c hints attribute is not supported.

  \sa ScXMLEventTarget, ScXMLEventElt, ScXMLDataElt, ScXMLAssignElt
  \ingroup scxml
*/

#include <cassert>
#include <cstring>
#include <string>
#include <vector>

#include <Inventor/errors/SoDebugError.h>
#include <Inventor/C/XML/element.h>
#include <Inventor/scxml/ScXMLEvent.h>
#include <Inventor/scxml/ScXMLStateMachine.h>
#include <Inventor/scxml/ScXMLEvaluator.h>

#include "scxml/ScXMLCommonP.h"
#include "SbBasicP.h"

// *************************************************************************

class ScXMLSendEltReader : public ScXMLEltReader {
public:
  ScXMLSendEltReader(void);
  virtual ScXMLElt * read(ScXMLElt * container, cc_xml_elt * xmlelt, ScXMLDocument * doc, ScXMLStateMachine * sm);
};

ScXMLSendEltReader::ScXMLSendEltReader(void)
: ScXMLEltReader("send")
{
}

ScXMLElt *
ScXMLSendEltReader::read(ScXMLElt * container, cc_xml_elt * xmlelt, ScXMLDocument * COIN_UNUSED_ARG(doc), ScXMLStateMachine * COIN_UNUSED_ARG(sm))
{
  assert(container && xmlelt);
  ScXMLSendElt * send = new ScXMLSendElt;
  send->setContainer(container);
  this->setXMLAttributes(send, xmlelt);

  // handle XML attributes
  if (unlikely(!send->handleXMLAttributes())) {
    delete send;
    return nullptr;
  }

  const int numchildren = cc_xml_elt_get_num_children(xmlelt);
  for (int c = 0; c < numchildren; ++c) {
    cc_xml_elt * element = cc_xml_elt_get_child(xmlelt, c);
    const char * elementtype = cc_xml_elt_get_type(element);

    if (std::strcmp(elementtype, COIN_XML_CDATA_TYPE) == 0) {
      // ignore CDATA
      continue;
    }

    SoDebugError::post("ScXMLSendEltReader::read",
                       "<send> contains unexpected <%s> element", elementtype);
    delete send;
    return nullptr;
   }

  return send;
}

// *************************************************************************

class ScXMLSendElt::PImpl {
public:
};

#define PRIVATE(obj) ((obj)->pimpl)

SCXML_ELEMENT_SOURCE(ScXMLSendElt);

void
ScXMLSendElt::initClass(void)
{
  SCXML_OBJECT_INIT_CLASS(ScXMLSendElt, ScXMLExecutableElt, "ScXMLExecutableElt");
  SCXML_ELEMENT_REGISTER_READER(ScXMLSendElt, "send", ScXMLSendEltReader);
}

void
ScXMLSendElt::cleanClass(void)
{
  SCXML_ELEMENT_UNREGISTER_READER(ScXMLSendElt);
  ScXMLSendElt::classTypeId = SoType::badType();
}

ScXMLSendElt::ScXMLSendElt(void)
: event(nullptr),
  target(nullptr),
  targettype(nullptr),
  sendid(nullptr),
  delay(nullptr),
  namelist(nullptr),
  hints(nullptr)
{
}

ScXMLSendElt::~ScXMLSendElt(void)
{
  this->setEventAttribute(nullptr);
  this->setTargetAttribute(nullptr);
  this->setTargetTypeAttribute(nullptr);
  this->setSendIDAttribute(nullptr);
  this->setDelayAttribute(nullptr);
  this->setNameListAttribute(nullptr);
  this->setHintsAttribute(nullptr);
}

void
ScXMLSendElt::setEventAttribute(const char * eventstr)
{
  SCXML__SET_ATTRIBUTE_VALUE(this->event, "event", eventstr);
}

// const char * getEventAttribute(void) const;

void
ScXMLSendElt::setTargetAttribute(const char * targetstr)
{
  SCXML__SET_ATTRIBUTE_VALUE(this->target, "target", targetstr);
}

// const char * getTargetAttribute(void) const;

void
ScXMLSendElt::setTargetTypeAttribute(const char * targettypestr)
{
  SCXML__SET_ATTRIBUTE_VALUE(this->targettype, "targettype", targettypestr);
}

// const char * getTargetTypeAttribute(void) const;

void
ScXMLSendElt::setSendIDAttribute(const char * sendidstr)
{
  SCXML__SET_ATTRIBUTE_VALUE(this->sendid, "sendid", sendidstr);
}

// const char * getSendIDAttribute(void) const;

void
ScXMLSendElt::setDelayAttribute(const char * delaystr)
{
  SCXML__SET_ATTRIBUTE_VALUE(this->delay, "delay", delaystr);
}

// const char * getDelayAttribute(void) const;

void
ScXMLSendElt::setNameListAttribute(const char * nameliststr)
{
  SCXML__SET_ATTRIBUTE_VALUE(this->namelist, "namelist", nameliststr);
}

// const char * getNameListAttribute(void) const;

void
ScXMLSendElt::setHintsAttribute(const char * hintsstr)
{
  SCXML__SET_ATTRIBUTE_VALUE(this->hints, "hints", hintsstr);
}

// const char * getHintsAttribute(void) const;

bool
ScXMLSendElt::handleXMLAttributes(void)
{
  if (!inherited::handleXMLAttributes()) {
    return false;
  }

  this->setEventAttribute(this->getXMLAttribute("event"));
  this->setTargetAttribute(this->getXMLAttribute("target"));
  this->setTargetTypeAttribute(this->getXMLAttribute("targettype"));
  this->setSendIDAttribute(this->getXMLAttribute("sendid"));
  this->setDelayAttribute(this->getXMLAttribute("delay"));
  this->setNameListAttribute(this->getXMLAttribute("namelist"));
  this->setHintsAttribute(this->getXMLAttribute("hints"));

  return true;
}

void
ScXMLSendElt::copyContents(const ScXMLElt * rhs)
{
  inherited::copyContents(rhs);
  const ScXMLSendElt * orig = coin_assert_cast<const ScXMLSendElt *>(rhs);
  this->setEventAttribute(orig->getEventAttribute());
  this->setTargetAttribute(orig->getTargetAttribute());
  this->setTargetTypeAttribute(orig->getTargetTypeAttribute());
  this->setSendIDAttribute(orig->getSendIDAttribute());
  this->setDelayAttribute(orig->getDelayAttribute());
  this->setNameListAttribute(orig->getNameListAttribute());
  this->setHintsAttribute(orig->getHintsAttribute());
}

const ScXMLElt *
ScXMLSendElt::search(const char * attrname, const char * attrvalue) const
{
  const ScXMLElt * hit = inherited::search(attrname, attrvalue);
  if (hit) {
    return hit;
  }
  if (std::strcmp(attrname, "event") == 0) {
    if (this->target && std::strcmp(attrvalue, this->event) == 0) {
      return this;
    }
  }
  else if (std::strcmp(attrname, "target") == 0) {
    if (this->target && std::strcmp(attrvalue, this->target) == 0) {
      return this;
    }
  }
  else if (std::strcmp(attrname, "targettype") == 0) {
    if (this->targettype && std::strcmp(attrvalue, this->targettype) == 0) {
      return this;
    }
  }
  else if (std::strcmp(attrname, "sendid") == 0) {
    if (this->sendid && std::strcmp(attrvalue, this->sendid) == 0) {
      return this;
    }
  }
  else if (std::strcmp(attrname, "delay") == 0) {
    if (this->delay && std::strcmp(attrvalue, this->delay) == 0) {
      return this;
    }
  }
  else if (std::strcmp(attrname, "namelist") == 0) {
    if (this->namelist && std::strcmp(attrvalue, this->namelist) == 0) {
      return this;
    }
  }
  else if (std::strcmp(attrname, "hints") == 0) {
    if (this->hints && std::strcmp(attrvalue, this->hints) == 0) {
      return this;
    }
  }
  return nullptr;
}

namespace {

void tokenize(const std::string & input, const std::string & delimiters, std::vector<std::string> & tokens, int count = -1)
{
  std::string::size_type last_pos = 0, pos = 0;
  while (true) {
    --count;
    pos = input.find_first_of(delimiters, last_pos);
    if ((pos == std::string::npos) || (count == 0)) {
      tokens.push_back(input.substr(last_pos));
      break;
    } else {
      tokens.push_back(input.substr(last_pos, pos - last_pos));
      last_pos = pos + 1;
    }
  }
}

} // namespace

ScXMLEvent *
ScXMLSendElt::createEvent(ScXMLEventTarget * host) const
{
  ScXMLEvent * event = new ScXMLEvent;

  if (!this->event) {
    // for now, inline xml-content is not supported, so we emit error.fetch for
    // that case
    event->setEventName("error.Fetch");
    host->queueInternalEvent(event);
    delete event;
    return nullptr;
  }

  event->setEventName(this->event);

  if (host->isOfType(ScXMLStateMachine::getClassTypeId())) {
    ScXMLStateMachine * sm = static_cast<ScXMLStateMachine *>(host);
    // standard associations
    event->setAssociation("_sessionid", sm->getSessionId().getString());
  }

  if (this->namelist &&
      (strlen(this->namelist) > 0) &&
      host->isOfType(ScXMLStateMachine::getClassTypeId()))
  {
    ScXMLStateMachine * sm = static_cast<ScXMLStateMachine *>(host);

    // user-specific associations
    std::string nameliststr = this->namelist;
    std::vector<std::string> tokens;
    tokenize(this->namelist, " ", tokens);

    // FIXME: use evaluator
    for (size_t c = 0; c < tokens.size(); ++c) {
      std::string token = tokens[c];
      const char * value = sm->getVariable(token.c_str());
      if (value) {
        event->setAssociation(token.c_str(), value);
      } else {
        ScXMLEvaluator * evaluator = sm->getEvaluator();
        if (evaluator) {
          ScXMLDataObj * dataobj = evaluator->locate(token.c_str());
          if (dataobj &&
              dataobj->getTypeId().isDerivedFrom(ScXMLConstantDataObj::getClassTypeId())) {
            SbString str;
            ScXMLConstantDataObj * cobj = static_cast<ScXMLConstantDataObj *>(dataobj);
            cobj->convertToString(str);
            event->setAssociation(token.c_str(), str.getString());
          } else {
            host->queueInternalEvent("error.send.InvalidNameListObject");
          }
        }
      }
    }
  }

  return event;
}

void
ScXMLSendElt::execute(ScXMLStateMachine * statemachine) const
{
  inherited::execute(statemachine);
  statemachine->sendExternalEvent(this);
}

#undef PRIVATE
